### PR TITLE
Forbid HTTP protocols other than 1.

### LINKED ISSUE #281: Forbid HTTP versions we don't understand.
Right now, NIO's sample server (and any server using NIO's HTTP layer) will be wildly over generous with the parsing in two different ways.

Firstly, it will accept *later* major versions of HTTP (e.g. a hypothetical HTTP/2.0, distinct from HTTP/2). This is a bad idea, because we don't know the semantics of those HTTP versions. We should treat those as parser errors.

Secondly, and more annoyingly, we will parse HTTP/0.9 requests, but don't know how to emit them. That means that incautious users of NIO may emit invalid HTTP responses that use the full-response format but claim a HTTP/0.9 version. Given that HTTP/0.9 is stupid, and given also that [http_parser doesn't even parse 0.9 properly](https://github.com/nodejs/http-parser/issues/386), we should not support HTTP/0.9 request or response formats, and treat them as errors.

Taken as a whole, this suggests a policy: NIOHTTP1 should enforce its own naming convention, and refuse to decode any HTTP message claiming a major version number other than 1. I don't propose to change the encoders, because users may want to have an escape hatch, but the decoders should refuse to emit such a message.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

Our HTTP code handles only HTTP/1.X. There is no reason to support
HTTP/0.9, and we cannot safely handle a major protocol higher than 1 in
this code, so we should simply treat requests/responses claiming to be
of those protocols as errors.

Modifications:

HTTPDecoder now checks the major version is equal to 1 before it
continues with parsing. If it hits an error, that error will be propagated
out to the user.

Result:

Better resilience against bad HTTP messages.

Resolves #281

### CODE REVIEW

File: Sources/NIOHTTP1/HTTPDecoder.swift [hunk 1]
```diff
@@ -25,6 +25,7 @@ private struct HTTPParserState {
     var readerIndexAdjustment = 0
     // This is set before http_parser_execute(...) is called and set to nil again after it finish
     var baseAddress: UnsafePointer<UInt8>?
+    var currentError: Error?
```
Comment (weissi) [inline Sources/NIOHTTP1/HTTPDecoder.swift]: you can save an existential here if you make the type `HTTPParserError`. That probably also save an allocation, maybe ;)

File: Tests/NIOHTTP1Tests/HTTPDecoderTest.swift [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        }
+
+        loop.run()
+    }
+
+    func testToleratesHTTP13Request() throws {
+        XCTAssertNoThrow(try channel.pipeline.add(handler: HTTPRequestDecoder()).wait())
+
+        // We tolerate higher versions of HTTP/1 than we know about because RFC 7230
+        // says that these should be treated like HTTP/1.1 by our users.
```
Comment (weissi) [inline Tests/NIOHTTP1Tests/HTTPDecoderTest.swift]: omg, what's wrong with RFC7230?

File: Sources/NIOHTTP1/HTTPDecoder.swift (hunk 1 above)
Comment (weissi) [reply Sources/NIOHTTP1/HTTPDecoder.swift]: actually, only saves an allocation on assignment. But definitely saves a couple of words in this structure if we don't make that an existential type

File: Tests/NIOHTTP1Tests/HTTPDecoderTest.swift (hunk 2 above)
Comment (Lukasa) [reply Tests/NIOHTTP1Tests/HTTPDecoderTest.swift]: I mean, the right thing for our users to do is to send HTTP/1.1 responses in response to HTTP/1.2 requests. But we can't really enforce that sensibly.

Comment (Lukasa) [reply Tests/NIOHTTP1Tests/HTTPDecoderTest.swift]: Anyway it's largely theoretical as HTTP/1.2 is never gonna happen.

File: Sources/NIOHTTP1/HTTPDecoder.swift (hunk 1 above)
Comment (Lukasa) [reply Sources/NIOHTTP1/HTTPDecoder.swift]: ACK, addressed.

**Comment by tomerd** [comment]
@swift-nio-bot test this please

**Comment by Lukasa** [comment]
Gah, this hit `testWeDontCrashIfChannelReleasesBeforePipeline` flakiness. I think there is a real timing issue here, probably best addressed by just sleeping the main thread.

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Pull request merged**